Automatically route PRs to specific remotes

abigailfox opened the pull request:
#### 6dffcfd53038053f56e81da06bb8941ea04435fd
<pre>
Automatically route PRs to specific remotes
<a href="https://bugs.webkit.org/show_bug.cgi?id=242129">https://bugs.webkit.org/show_bug.cgi?id=242129</a>
rdar://96161931

Reviewed by Jonathan Bedard.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request):

Canonical link: <a href="https://commits.webkit.org/252715@main">https://commits.webkit.org/252715@main</a>
</pre>

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -400,6 +419,8 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
             print("Updated '{}'!".format(pr))
         else:
+            # TODO: I NEED TO BE REMOVED LATER
+            return 1
```
This can be removed now, yes?

abigailfox replied:
I just caught these from the style checker, they should be gone now

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -331,8 +331,34 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+                break
+
         if isinstance(remote_repo, remote.GitHub):
+            if issue and issue.redacted and args.remote is None: # TODO: is this a correct check on args.remote?
```
Remove the to-do here.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -331,8 +331,34 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+
         if isinstance(remote_repo, remote.GitHub):
+            if issue and issue.redacted and args.remote is None: # TODO: is this a correct check on args.remote?
+                print("Your issue is redacted, which means it will be diverted to a secure, non-origin remote you have access to.")
```
Would drop the "which means it will" in exchange for just "diverting"

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -331,8 +331,34 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+                ) == 'No':
+                    original_remote = source_remote
+                    if len(repository.source_remotes()) < 2:
+                        print("Error. You do not have access to a secure, non-origin remote")
```
Should send this to `sys.stderr` instead of printing it.

abigailfox replied:
Should this one go to stderr, or the following one that actually cancels it? This one prompts them to reconsider overriding and then they can select

JonWBedard replied:
This one. Basically, the stuff that explains why you're returning an error code.

abigailfox replied:
Should I just not even ask them if they want to proceed anyways? Just go directly to stderr, no further Terminal interaction?

JonWBedard replied:
Oh, I'm sorry, I misread this. I still think this should be `sys.stderr`, since you're starting the line with `error`, but it's not as clear cut as I originally said, sorry.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -331,8 +331,34 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+                            print("Failed to create pull request due to non-suitable remote")
+                            return 1
+                    source_remote = repository.source_remotes()[1]
+                    print("Switched remote from {} to {}".format(original_remote, source_remote))
```
Would say `Making PR against '{}' instead of '{}'".format(source_remote, original_remote)`

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -331,8 +331,34 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+                        return 1
+                    else:
+                        target = 'fork' if source_remote == repository.default_remote else '{}-fork'.format(
+                            source_remote)
```
Nit: You don't need this `else` statement at all since this same statement is run unconditionally on line 360

JonWBedard reviewed (commented):
Should probably mention `Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py` in the commit message, even if not the specific function you're modifying.

abigailfox committed:
Automatically route PRs to specific remotes
https://bugs.webkit.org/show_bug.cgi?id=242129
rdar://96161931

Reviewed by Jonathan Bedard.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request):

Canonical link: https://commits.webkit.org/252715@main
